zebra-network/src/peer_set/candidate_set/tests/prop.rs

```diff
@@ -17,7 +17,60 @@ proptest! {
+            prop_assert!(peer.get_last_seen().timestamp() <= last_seen_limit.timestamp() + 1,
+                         "peer timestamp {} was greater than limit {}",
+                         peer.get_last_seen().timestamp(),
+                         last_seen_limit.timestamp());
+
+            /* These tests won't pass until PR #2203 is merged
+
+            use zebra_chain::serialization::{ZcashDeserialize, ZcashSerialize};
+
+            // Check that malicious peers can't make Zebra send bad times to other peers
+            // (after Zebra's standard sanitization)
+            let sanitized_peer = peer.sanitize();
+
+            // Check that sanitization doesn't put times in the future
+            prop_assert!(sanitized_peer.get_last_seen().timestamp() <= last_seen_limit.timestamp() + 1,
+                         "sanitized peer timestamp {} was greater than limit {}, original timestamp: {}",
+                         sanitized_peer.get_last_seen().timestamp(),
+                         last_seen_limit.timestamp(),
+                         peer.get_last_seen().timestamp());
+
+            // Check that malicious peers can't make Zebra's serialization fail
+            let addr_bytes = peer.zcash_serialize_to_vec();
+            prop_assert!(addr_bytes.is_ok(),
+                         "unexpected serialization error: {:?}, original timestamp: {}, sanitized timestamp: {}",
+                         addr_bytes,
+                         peer.get_last_seen().timestamp(),
+                         sanitized_peer.get_last_seen().timestamp());
+
+            // Assume other implementations deserialize like Zebra
+            let deserialized_peer = MetaAddr::zcash_deserialize(addr_bytes.unwrap().as_slice());
+            prop_assert!(deserialized_peer.is_ok(),
+                         "unexpected deserialization error: {:?}, original timestamp: {}, sanitized timestamp: {}",
+                         deserialized_peer,
+                         peer.get_last_seen().timestamp(),
+                         sanitized_peer.get_last_seen().timestamp());
+            let deserialized_peer = deserialized_peer.unwrap();
+
+            // Check that serialization hasn't modified the address
+            // (like the sanitized round-trip test)
+            prop_assert_eq!(sanitized_peer, deserialized_peer);
+
+            // Check that sanitization, serialization, and deserialization don't
+            // put times in the future
+            prop_assert!(deserialized_peer.get_last_seen().timestamp() <= last_seen_limit.timestamp() + 1,
+                         "deserialized peer timestamp {} was greater than limit {}, original timestamp: {}, sanitized timestamp: {}",
+                         deserialized_peer.get_last_seen().timestamp(),
+                         last_seen_limit.timestamp(),
+                         peer.get_last_seen().timestamp(),
+                         sanitized_peer.get_last_seen().timestamp());
```

Comment:
I think I'm also missing something about these tests. Should each of them be separate `#[proptest]`s?

Also, should some of them be done in separate PRs? I mean, I don't understand how serialization and sanitization are related to the code to ensure the time limit. My intuition is that there should be a `proptest` to ensure that any arbitrary `MetaAddr` is serializable correctly and can be properly sanitized?

Reply:
I'm still working through all of this myself.

Here's where I've got to today:

If we want to serialize the last seen time correctly, it needs to fit in a `u32`. But `DateTime<Utc>` returns seconds in `i64`. (And uses about ~56 bits internally.) So it might not uphold this invariant.

> My intuition is that there should be a `proptest` to ensure that any arbitrary `MetaAddr` is serializable correctly and can be properly sanitized?

There is now. I wrote those tests in #2203 and they just got committed as 9f8b4f836ec8d7a039383eb80ba4a6369411f9cf and 5cdcc5255f8c66fe0fc0f2a41b6e7b37e8530b1f.

Sanitization truncates to the nearest 30 minutes. Since timestamp `0` is already at a 30 minute mark, it just happens to uphold the `u32` range invariant. (Despite `DateTime<Utc>` not requiring that invariant.)

> Also, should some of them be done in separate PRs? I mean, I don't understand how serialization and sanitization are related to the code to ensure the time limit. 

The validation we're adding subtracts arbitrary times. This can move the last seen time outside the `u32` range. So the `MetaAddr` appears to be valid, until we go to sanitize and serialize it. Then the entire connection gets closed due to the serialization error. (Which is a denial of service risk.)

I used these proptests to discover that error. (And similar errors with leap seconds.)

I think the best way to fix these bugs is to create a `DateTime32(u32)` type. (Which is part of ticket #2171.) Using a 32-bit internal representation will make sure we can't go outside that range.

Otherwise we will have to validate 32-bit datetimes every time we use or modify them. (In general, we try to use representations that automatically ensure our constraints.)

Reply:
I'll go work on `DateTime32` now.

Reply:
I created `DateTime32` and made `MetaAddr` use it in https://github.com/ZcashFoundation/zebra/pull/2210

For the `CandidateSet` validation, here's how we could move forward:
* take the limit as `DateTime32`
    * panic if the current time is outside of `DateTime32` (by 2038 the protocol should have been updated)
* do all the calculations in `chrono::DateTime` (better API, no overflows or underflows because it's ~56 bits)
* convert the times to `DateTime32` for each address, returning an error if they're out of range